src/mono/System.Private.CoreLib/src/System/Exception.Mono.cs

```diff
@@ -91,9 +106,14 @@ internal DispatchState CaptureDispatchState()
 
         internal void RestoreDispatchState(in DispatchState state)
         {
-            foreignExceptionsFrames = state.StackFrames;
-
-            _stackTraceString = null;
+            // Isolate so we can safely update foreignExceptionFrames and CaptureDispatchState can read the correct values
+            //
+            // See https://github.com/dotnet/runtime/issues/70081
+            lock(frameLock)
+            {
+                foreignExceptionsFrames = state.StackFrames;
```

Comment:
From what I can see, I don't believe `state.StackFrames` can be null.  `Diagnostics.StackTrace.get_trace(this, 0, true)` is an icall to a native function that appears to always return an array.

@lambdageek - can you please help confirm?

Reply:
yes, AFAICT `StackTrace.get_trace` always returns an array or throws an exception. It does not return `null`

Reply:
get_trace is only called if _traceIPs is not null.  In what situation is _traceIPs null?  Seems like it could be null if the exception wasn't thrown?  So what happens if a new Exception is created, the dispatch info is captured before it's thrown, then it's thrown, then it's captured again, and now there's a race between those two different dispatch infos being restored?

Reply:
It looks like there is some magic integration between `Exception.Mono.cs` and native code. It's likely there are gaps in my understanding, but I'll try my best to answer your question.

I believe `_traceIPs` becomes not null only after an exception instance is thrown.  In your example, the dispatch info and `foreignExceptionFrames` would continue to be null until the exception is thrown.  If there was a concurrent capture and restore, I think `foreignExceptionFrames` would continue to be null no matter what until the first `get_trace` read and subsequent restore.  

I guess what could happen is the stack frame could be flagged with `isLastFrameFromForeignException` == true and `foreignExceptionFrames` could still be null if a concurrent Capture / Restore happens and there is no follow up Restore. It's unclear if this matters at all.

@lambdageek @marek-safar your thoughts?

Reply:
Yea, `_traceIPs` is set from native code in mono-exceptions.c after unwinding.

Capture/Restore are pretty rare, right? Can we just lock everything in them?